### PR TITLE
Encapsed literal strings

### PR DESCRIPTION
**Pull request by AndrolGenhald**
- Infer literal string from encapsed string.
 - Allow casting literal float to literal string
 - Improve concatenation of literals to allow literal floats, avoid long loop if there are too many combinations instead of checking after loop
 - Improve `AssertionFinder::hasNonEmptyCountEqualityCheck()` and add tests for #8163

### CODE REVIEW

File: src/Psalm/Internal/Analyzer/Statements/Expression/AssertionFinder.php [hunk 1]
```diff
@@ -1499,50 +1501,35 @@ protected static function hasNonEmptyCountEqualityCheck(
...
+            return false;
+        }
 
-        if ($right_count
-            && $conditional->left instanceof PhpParser\Node\Scalar\LNumber
-            && $operator_less_than_or_equal
-            && $conditional->left->value >= (
-                $conditional instanceof PhpParser\Node\Expr\BinaryOp\Smaller ? 0 : 1
-            )
+        // TODO get node type provider here somehow and check literal ints and int ranges
```
Comment (orklah) [inline src/Psalm/Internal/Analyzer/Statements/Expression/AssertionFinder.php:1526]: you can pass $source from caller method and then do `$source->getNodeTypeProvider()` if you want

Comment (AndrolGenhald) [reply src/Psalm/Internal/Analyzer/Statements/Expression/AssertionFinder.php:1526]: I can't believe I missed that, I was looking for a `StatementsAnalyzer`... I'm about done for now, but I'll come fix that in the next day or so.

Comment (orklah) [reply src/Psalm/Internal/Analyzer/Statements/Expression/AssertionFinder.php:1526]: otherwise, the rest seems good :) Give me a sign when it's ready ;)

Comment (AndrolGenhald) [reply src/Psalm/Internal/Analyzer/Statements/Expression/AssertionFinder.php:1526]: Ok, well this is actually a bit complicated. I got it all working, but I got a couple `RedundantCondition`s... I have to figure out a way to make the assertion not redundant if `$foo` and `$bar` are both non-empty and there's an assertion for `count($foo) === count($bar)`.

Go ahead and merge this, I'll have to dig deeper later. I'll probably see if I can fix this on master along with #7096.

**Comment by orklah** [comment]
Thanks!

**Pull request merged**